src/nvim/extmark.c

```diff
@@ -112,6 +112,7 @@ void extmark_set(buf_T *buf, uint32_t ns_id, uint32_t *idp, int row, colnr_T col
           marktree_revise(buf->b_marktree, itr, decor_level, old_mark);
           goto revised;
         }
+        decor_remove(buf, old_mark.pos.row, old_mark.pos.row, old_mark.decor_full);
```

Comment:
This could also have the `if (marktree_decor_level(old_mark) > kDecorLevelNone)` check as above.

Reply:
> I considered that, but it seemed that decor_remove gates on decor_full being valid, and I didn't want to risk leaking anything.

If the there is no decoration there is nothing to leak.

> The call to decor_remove in extmark_clear also doesn't have this guard. Should it?

It does already, but more indirectly.

Reply:
You might need to be more forthcoming. I don't see what the effect would be of adding that extra code. If the decor is there it will be freed, otherwise not. Will it be more cache efficient or something?

Reply:
it avoids an extra redraw if there is no decoration which was removed.

Reply:
Would it not make more sense just to gate that specific redraw within `decor_remove`?

Reply:
Sounds like a refactor. We all want to refactor decoration code to be more flexible and efficient later :) This is just a missing edge case that currently makes this `decor_remove` call needlessly different than all other decor_remove calls (and will make the code harder to reason about if this refactor happens in 6 months and we all forgor about this)